### DISCUSSION TITLE
Poor performance on my own dataset

### DISCUSSION THREAD

**Discussion by qhdqhd** [NVlabs/instant-ngp#260] [Q&A]
My dataset contains about 100 views, the viewpoints form a ring, all viewpoints are looking towards the circle center. 
The cameras are like this:
 
![FQ)1(4{OMH%W`6G 9Q)T4%E](https://user-images.githubusercontent.com/42370719/155328629-937cff75-cc4d-4775-ac7e-1e5e373387ea.png)

The scene size is about 15 meters. 

The first row shows 5 rendered views of the training view. The second row shows 5 training views.

My parameter:
aabb_scale is set to 16 and I change the scale or offset in the camera positions as  https://github.com/NVlabs/instant-ngp/blob/master/docs/nerf_dataset_tips.md, so the scene has been fully covered by the box.

The people in the center of the scene renders fine, but there are quite a few white floating objects or ghosting around the people.
I would like to ask why?
There are three possible reasons I can think of:
1. Is it because the scene has too few textures? Lots of black backgrounds around.
2. Or is it due to camera parameters? Because I used 100 different cameras here, but the cameras are all 12mm, and the intrinsic parameters are very close, but not exactly the same. Since your project does not support multiple camera intrinsics, only one common intrinsic is used.
3. Or the camera shooting position? My cameras are all on a plane without views from different directions like looking down and looking up.

**Comment by Tom94** [comment]
Hi there, first of call: cool setup!

(1) shouldn't be a problem as long as the camera poses are precise enough for the `Train extrinsics` option to fix any slight inaccuracies.

I suspect the problem comes more from (2) and (3).

For (2), we recently (yesterday) pushed support for per-camera metadata. You can customize it via the python bindings
```python
testbed.nerf.training.dataset.metadata[image_id].camera_distortion = ...
testbed.nerf.training.dataset.metadata[image_id].focal_length = ...
testbed.nerf.training.dataset.metadata[image_id].principal_point = ...
```
More specifics are in `python_api.cu`. The `.json`-based loader unfortunately only supports a single set of camera parameters per json file.

For (3), this will largely manifest as artifacts when trying to view the scene from the top of bottom (i.e. outside of the convex hull of the training data). If you plan for the viewpoint to stay close to the ring of cameras, you should be fine.

Curious to hear whether you find the Python bindings w.r.t. (2) helpful.

Cheers!

**Reply by qhdqhd** [reply]
thank you for your fast anwser!
I still have a question. If the calibration accuracy is good enough, does the texture of the scene have a great impact on the training? Rich texture should be better than low texture, right?

**Reply by Tom94** [reply]
Rich texture is indeed better, yes, but less important for training than for pose estimation.

**Comment by slash-under** [comment]
In my experience, the "smoke" you're describing is most likely loss, which would disappear with more training data/potentially more points extracted with COLMAP.

I've tweaked some parameters in my fork, maybe applying the changes from here could help?

slash-under/instant-ngp@a8a3d76

**Reply by qhdqhd** [reply]
Your improvement helps to get better calibration accuracy by colmap, but my data seems to have a good calibration accuracy. Do you mean the "smoke" is due to inaccurate calibration?

**Reply by slash-under** [reply]
I believe it's inaccuracy (loss) from all stages of the process, the only
way I've found to lessen it further is by tweaking the angles of the input
dataset (sometimes by using a different network to create frames
in-between) or increasing one of the various quality "knobs", per se, like
increasing the total number of neurons.

**Reply by slash-under** [reply]
As in my example, that "smoke" you're describing could very well be missing points, as the background is generally black.

https://user-images.githubusercontent.com/63025235/156563919-5fef7b16-d0d1-4235-a7dc-58dbb37a9fb7.mp4

**Reply by qhdqhd** [reply]
Thank you! But it still confuses me. Do you mean a solid colored background is more prone to generate smoke?

**Reply by slash-under** [reply]
No, but the amount is heavily dependent on the input resolution. Above was trained on 250 4K frames, and outputs roughly 1080p quality under ideal conditions, and about 720p quality under most conditions. 


If the result quality shows a notable dependence on the view angle, some of the frames may not have had features extracted correctly.

**Reply by slash-under** [reply]
A better way to view it may be that the current process requires input data at about 2-4x the resolution of the desired output resolution for best quality.

**Reply by qhdqhd** [reply]
Thank you. My input views are also 4K, and output rendering views are 1080p.

**Reply by slash-under** [reply]
Does the "smoke" resolve if you move to a different view represented in the training data? It sounds like there might not be enough extracted from some of the frames.

**Reply by qhdqhd** [reply]
Yes,  just a small part of training views have "smoke", not all.

**Reply by slash-under** [reply]
It might help to append a portion of video from the angles with the most loss to the dataset, in that case.

**Reply by slash-under** [reply]
(or one could look through here for more tweaks I've missed: https://colmap.github.io/faq.html)

**Reply by slash-under** [reply]
If you are using the run.py script to extract frames from images, it appears that the use of JPEG format for the frames is introducing some compression error. I've changed it to PNG in my fork, and am awaiting results to see if the output quality changes.

**Reply by qhdqhd** [reply]
Thank you, I generally use JPEG. But does this really matter?

**Reply by HumbleShaw** [reply]
I'm so sorry, but is there change in your improvement ? @slash-under
